Add additional user profile attributes from LDAP

slapcat opened the pull request:
* Resolves: #40194 

## Summary
- Extend user_ldap User Profile Attributes to include birthday and anniversary.
- Update CardDAV converter to include these attributes in contacts for the Accounts address book.
- Add required constants to AccountManager.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes:
![before](https://github.com/nextcloud/server/assets/52802566/61aadcd7-7f2e-4806-80e6-4114f207f2b4)
![after](https://github.com/nextcloud/server/assets/52802566/0617557a-62e2-4779-b875-1ce126b7d73e)


- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
**EDIT:** Documentation PR [#11323](https://github.com/nextcloud/documentation/pull/11323)

- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

ChristophWurst commented on apps/dav/lib/CardDAV/Converter.php:
```diff
@@ -134,6 +134,15 @@ public function createCardFromUser(IUser $user): ?VCard {
 					$vCard->add(new Text($vCard, 'TITLE', $property->getValue(), ['X-NC-SCOPE' => $scope]));
 					break;
+				case IAccountManager::PROPERTY_BIRTHDATE:
+					$vCard->add(new Text($vCard, 'BDAY;VALUE=DATE', substr($property->getValue(), 0, 8), ['X-NC-SCOPE' => $scope]));
```
what value will $scope have?

slapcat replied:
In my test environment, $scope is `v2-local`. I only included it to match how the previous properties get set. It can be removed if not required.

ChristophWurst replied:
I'm asking because it's not clear to me if this property is always fine to expose to other users. Users can not change the scope and influence who sees their birth date.

slapcat replied:
Since they are not necessary for this feature anyhow, I have removed the scope from those lines.

ChristophWurst replied:
No scope has the same semantic as scope private, so the property will only show for the user

kesselb commented on apps/dav/lib/CardDAV/Converter.php:
```diff
@@ -134,6 +134,15 @@ public function createCardFromUser(IUser $user): ?VCard {
+				case IAccountManager::PROPERTY_ANNIVERSARYDATE:
+					$vCard->add(new Text($vCard, 'ANNIVERSARY;VALUE=DATE', substr($property->getValue(), 0, 8), ['X-NC-SCOPE' => $scope]));
+					/* iOS compatibility */
+					$vCard->add(new Text($vCard, 'ITEM3.X-ABDATE;TYPE=pref', substr($property->getValue(), 0, 8), ['X-NC-SCOPE' => $scope]));
```
This "ITEM3." thing is a bit tricky: https://github.com/nextcloud/contacts/issues/223

slapcat replied:
IIUC, this issue and another like it I found (https://github.com/nextcloud/contacts/issues/2560) concern importing vCards with these ITEMx keys into NC and displaying them in the profile. This is not an issue regarding the LDAP => NC sync because the `ANNIVERSARY` key is also added, which NC understands and displays in the profile. I don't have an Android, but when I export from NC and import to Google Contacts in the browser, the anniversary date displays correctly. If Android *requires* the `X-ANDROID-CUSTOM` key to display the anniversary, I can add this also to the code.

kesselb replied:
I would like to know if it still works on iOS if you drop `ITEM3.X-ABDATE`? 

ANNIVERSARY is the only attribute we should write: https://www.rfc-editor.org/rfc/rfc6350.html#section-6.2.6

Everything else is off standard.

slapcat replied:
It does not work on iOS at all without the `ITEM3.X-ABDATE`. When you create a contact on iOS natively, it uses this key as well instead of following the standard.

kesselb replied:
Hey @slapcat, thanks for checking! 

I guess we don't have much of a choice then ;) 

I created the attached vcard using the online version of icloud contacts. 
You can add many anniversary dates and thus the itemX prefix. 

As we only have one anniversary date right now, it should be fine to use item1.

![image](https://github.com/nextcloud/server/assets/3902676/fbb01b06-77d5-498d-ad3d-57898c83916f)


[Jane Doe.vcf.txt](https://github.com/nextcloud/server/files/13532083/Jane.Doe.vcf.txt)

slapcat committed:
Add additional user profile attributes from LDAP

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
remove extra spaces after line

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
Remove adding X-NC-SCOPE to bday/anniversary in Contact card

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
change item3 to item1

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
resolve other conflicts

Signed-off-by: Jake Nabasny <jake@nabasny.com>

kesselb commented:
Thank you :+1: 

Will try to give it a test next week. 

I see a few failing tests. 
The following patch should fix one of them. 

```
Index: tests/lib/Accounts/AccountManagerTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/lib/Accounts/AccountManagerTest.php b/tests/lib/Accounts/AccountManagerTest.php
--- a/tests/lib/Accounts/AccountManagerTest.php	(revision 7eaa573bac0f82b5b2b837104a508dab37c6e231)
+++ b/tests/lib/Accounts/AccountManagerTest.php	(date 1701547210963)
@@ -632,6 +632,18 @@
 				'scope' => IAccountManager::SCOPE_LOCAL,
 			],
 
+			[
+				'name' => IAccountManager::PROPERTY_BIRTHDATE,
+				'value' => '',
+				'scope' => IAccountManager::SCOPE_LOCAL,
+			],
+
+			[
+				'name' => IAccountManager::PROPERTY_ANNIVERSARYDATE,
+				'value' => '',
+				'scope' => IAccountManager::SCOPE_LOCAL,
+			],
+
 			[
 				'name' => IAccountManager::PROPERTY_PROFILE_ENABLED,
 				'value' => '1',

```

kesselb commented on apps/settings/lib/Settings/Personal/PersonalInfo.php:
```diff
@@ -167,6 +167,8 @@ public function getForm(): TemplateResponse {
 			'role' => $this->getProperty($account, IAccountManager::PROPERTY_ROLE),
 			'headline' => $this->getProperty($account, IAccountManager::PROPERTY_HEADLINE),
 			'biography' => $this->getProperty($account, IAccountManager::PROPERTY_BIOGRAPHY),
+			'birthDate' => $this->getProperty($account, IAccountManager::PROPERTY_BIRTHDATE),
```
Seems unused because the view was not updated to show the data in the user profile. 
I'd say we should remove it here for now.

kesselb commented:
This should fix the second failing tests by updating the expected data. 

```diff
Index: apps/settings/tests/UserMigration/assets/account.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/apps/settings/tests/UserMigration/assets/account.json b/apps/settings/tests/UserMigration/assets/account.json
--- a/apps/settings/tests/UserMigration/assets/account.json	(revision 7eaa573bac0f82b5b2b837104a508dab37c6e231)
+++ b/apps/settings/tests/UserMigration/assets/account.json	(date 1701548042464)
@@ -1,1 +1,0 @@
-{"displayname":{"name":"displayname","value":"Emma Jones","scope":"v2-federated","verified":"0","verificationData":""},"address":{"name":"address","value":"920 Grass St","scope":"v2-local","verified":"0","verificationData":""},"website":{"name":"website","value":"","scope":"v2-local","verified":"0","verificationData":""},"email":{"name":"email","value":"","scope":"v2-federated","verified":"0","verificationData":""},"avatar":{"name":"avatar","value":"","scope":"v2-federated","verified":"0","verificationData":""},"phone":{"name":"phone","value":"","scope":"v2-local","verified":"0","verificationData":""},"twitter":{"name":"twitter","value":"","scope":"v2-local","verified":"0","verificationData":""},"fediverse":{"name":"fediverse","value":"","scope":"v2-local","verified":"0","verificationData":""},"organisation":{"name":"organisation","value":"","scope":"v2-local","verified":"0","verificationData":""},"role":{"name":"role","value":"","scope":"v2-local","verified":"0","verificationData":""},"headline":{"name":"headline","value":"","scope":"v2-local","verified":"0","verificationData":""},"biography":{"name":"biography","value":"","scope":"v2-local","verified":"0","verificationData":""},"profile_enabled":{"name":"profile_enabled","value":"1","scope":"v2-local","verified":"0","verificationData":""},"additional_mail":[]}
\ No newline at end of file
Index: apps/settings/tests/UserMigration/assets/account-complex.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/apps/settings/tests/UserMigration/assets/account-complex.json b/apps/settings/tests/UserMigration/assets/account-complex.json
--- a/apps/settings/tests/UserMigration/assets/account-complex.json	(revision 7eaa573bac0f82b5b2b837104a508dab37c6e231)
+++ b/apps/settings/tests/UserMigration/assets/account-complex.json	(date 1701548067344)
@@ -1,1 +1,0 @@
-{"displayname":{"name":"displayname","value":"Steve Smith","scope":"v2-local","verified":"0","verificationData":""},"address":{"name":"address","value":"123 Water St","scope":"v2-local","verified":"0","verificationData":""},"website":{"name":"website","value":"https:\/\/example.org","scope":"v2-local","verified":"0","verificationData":""},"email":{"name":"email","value":"steve@example.org","scope":"v2-federated","verified":"0","verificationData":""},"avatar":{"name":"avatar","value":"","scope":"v2-local","verified":"0","verificationData":""},"phone":{"name":"phone","value":"+12178515387","scope":"v2-private","verified":"0","verificationData":""},"twitter":{"name":"twitter","value":"steve","scope":"v2-federated","verified":"0","verificationData":""},"fediverse":{"name":"fediverse","value":"@steve@floss.social","scope":"v2-federated","verified":"0","verificationData":""},"organisation":{"name":"organisation","value":"Mytery Machine","scope":"v2-private","verified":"0","verificationData":""},"role":{"name":"role","value":"Manager","scope":"v2-private","verified":"0","verificationData":""},"headline":{"name":"headline","value":"I am Steve","scope":"v2-local","verified":"0","verificationData":""},"biography":{"name":"biography","value":"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris porttitor ullamcorper dictum. Sed fermentum ut ligula scelerisque semper. Aliquam interdum convallis tellus eu dapibus. Integer in justo sollicitudin, hendrerit ligula sit amet, blandit sem.\n\nSuspendisse consectetur ultrices accumsan. Quisque sagittis bibendum lectus ut placerat. Mauris tincidunt ornare neque, et pulvinar tortor porttitor eu.","scope":"v2-local","verified":"0","verificationData":""},"profile_enabled":{"name":"profile_enabled","value":"1","scope":"v2-local","verified":"0","verificationData":""},"additional_mail":[{"name":"additional_mail","value":"steve@example.com","scope":"v2-published","verified":"0","verificationData":""},{"name":"additional_mail","value":"steve@earth.world","scope":"v2-local","verified":"0","verificationData":""}]}
\ No newline at end of file

```

slapcat committed:
Add new properties to AccountManagerTest
Remove properties from PersonalInfo, since not visible in profile

Signed-off-by: Jake Nabasny <jake@nabasny.com>

kesselb commented:
> Side note: I thought the `biography` field would be converted to the carddav `notes`, but see now it is not and has no correlated field for the contact card. @kesselb would it be okay if I add this to this PR, or is this kept separate for a reason?

Good catch, sounds good to me. 

@miaulalala @ChristophWurst mapping the about field from our profile, which should be populated from the ldap biography field if available, to the notes field in vcard. Wdyt?

ChristophWurst commented:
I think we should map LDAP attributes to profile properties, and profile properties to the system contact. Then users can set the scope of the birthday and anniversary fields.

Without a scope setting we either
1) Have to make the property private, which means nobody will see it so it's useless.
2) Have to make the property local, which means all other users can see your birthday, even if you don't want to expose this.

slapcat committed:
- Convert biography to carddav NOTE field
- Add back including the birthdate and anniversary in the profile

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
Add back scope as requested by @ChristophWurst

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat commented:
I added back the scoping as suggested to the code I already wrote. I'm not as familiar with the frontend and would find it difficult to add these features correctly to the profile page. Can someone assist with that portion? Or can we split that into a separate PR?

slapcat committed:
Fixing merged code

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
Undo partial UI changes

This reverts commit 459355f4e8d6f90a97598232d6ed25d050efbd65.

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
Clean up code merge markers

Signed-off-by: Jake Nabasny <jake@nabasny.com>

slapcat committed:
Fix Anniversary carddav label

Signed-off-by: Jake Nabasny <jake@nabasny.com>

st3iny commented:
Hey @slapcat, a colleague of mine and me hacked together a birthday field and added it to the profile a while ago. Perhaps the frontend code could be reused here to allow configuring the birthday and set the scope.

There is also working backend code for saving a user's birthday property.

The frontend is mostly in:
https://github.com/nextcloud/server/blob/feat/account/add-birthday-prop/apps/settings/src/components/PersonalInfo/BirthdaySection.vue

The full PR is at: https://github.com/nextcloud/server/pull/39592

st3iny commented:
The alternative PR including a documentation PR are merged.

Ref https://github.com/nextcloud/server/pull/45512
Ref https://github.com/nextcloud/documentation/pull/11323

slapcat commented:
Thanks @st3iny !
